### PR TITLE
fix(core/dfn-panel): precompute all dfnPanels

### PR DESCRIPTION
**Pull request by sidvishnoi**
This pre-computes all dfn panels and appends them to body. Then click listener only needs to hide/show and position already existing panels.

Precursor to https://github.com/w3c/respec/pull/2820

Tests have changed a lot because earlier we relied on existence of panel in DOM, but now we've to check if it's visible (`.hidden`) or not.

### CODE REVIEW

File: src/core/dfn-panel.js [hunk 1]
```diff
@@ -14,6 +14,13 @@ export async function run() {
     document.querySelector("link")
   );
 
+  /** @type {HTMLElement} */
+  const elems = document.querySelectorAll("dfn, .index-term");
+  for (const el of elems) {
+    const panel = createPanel(el);
+    document.body.append(panel);
```
Comment (sidvishnoi) [inline src/core/dfn-panel.js:21]: Would it be better to add all panels to a fragment and then append all at once?

File: src/core/dfn-panel.js [hunk 2]
```diff
@@ -69,9 +75,10 @@ function createPanel(dfn) {
   const href = dfn.dataset.href || `#${id}`;
   const links = document.querySelectorAll(`a[href="${href}"]:not(.index-term)`);
 
+  const panelId = `dfn-panel-for-${dfn.id}`;
   /** @type {HTMLElement} */
   const panel = hyperHTML`
-    <aside class="dfn-panel" id="dfn-panel">
+    <aside class="dfn-panel hidden removeOnSave" id="${panelId}">
```
Comment (sidvishnoi) [inline src/core/dfn-panel.js:81]: `removeOnSave` until https://github.com/w3c/respec/pull/2820

File: src/core/dfn-panel.js [hunk 3]
```diff
@@ -142,7 +149,10 @@ function getReferenceTitle(link) {
  * @param {{ x: number, y: number }} clickPosition
  */
 function displayPanel(dfn, panel, { x, y }) {
-  document.body.appendChild(panel);
+  for (const el of document.querySelectorAll(".dfn-panel")) {
```
Comment (sidvishnoi) [inline src/core/dfn-panel.js:152]: Maybe?
```suggestion
  for (const el of document.querySelectorAll(".dfn-panel:not(.hidden)")) {
```

File: src/ui/dfn-list.js [hunk 4]
```diff
@@ -78,7 +78,9 @@ function labelDfnIfExported(dfn) {
  * @param {HTMLElement} dfn a definition
  */
 function labelDfnIfUnused(dfn) {
-  const isUsed = document.querySelector(`a[href^="#${dfn.id}"]`);
+  const isUsed = document.querySelector(
+    `:not(.dfn-panel) > :not(li) > a[href^="#${dfn.id}"]`
```
Comment (sidvishnoi) [inline src/ui/dfn-list.js:82]: I'm not sure how this works, but it works :sweat_smile:

File: assets/dfn-panel.css [hunk 5]
```diff
@@ -84,3 +84,7 @@ dfn {
   max-height: 30vh;
   overflow: auto;
 }
+
+.dfn-panel.hidden {
```
Comment (saschanaz) [inline assets/dfn-panel.css:88]: There is an attribute named [`hidden`](https://html.spec.whatwg.org/multipage/interaction.html#the-hidden-attribute) which does the same 👀

File: src/core/dfn-panel.js (hunk 1 above)
Comment (saschanaz) [reply src/core/dfn-panel.js:21]: That might be cleaner.

File: assets/dfn-panel.css (hunk 5 above)
Comment (sidvishnoi) [reply assets/dfn-panel.css:88]: How can I forget!

**Comment by sidvishnoi** [comment]
> I personally prefer to only compute dfn maps and do anything else on runtime

Is this store references as JSON idea? If so, I'm not convinced how it may be used. A parser would need as much effort to parse `<aside>` than to find and parse a `<script type="application/json">`

File: src/core/dfn-panel.js [hunk 6]
```diff
@@ -49,29 +57,36 @@ function deriveAction(clickTarget) {
...
 function createPanel(dfn) {
   const { id } = dfn;
   const href = dfn.dataset.href || `#${id}`;
   const links = document.querySelectorAll(`a[href="${href}"]:not(.index-term)`);
 
+  const panelId = `dfn-panel-for-${dfn.id}`;
   /** @type {HTMLElement} */
   const panel = hyperHTML`
-    <aside class="dfn-panel" id="dfn-panel">
+    <aside class="dfn-panel removeOnSave" id="${panelId}" hidden>
```
Comment (marcoscaceres) [inline src/core/dfn-panel.js:89]: why remove?

Comment (sidvishnoi) [reply src/core/dfn-panel.js:89]: https://github.com/w3c/respec/pull/2901#discussion_r429569305

File: src/ui/dfn-list.js (hunk 4 above)
Comment (marcoscaceres) [reply src/ui/dfn-list.js:82]: Maybe add this 👆 as a comment? The query is indeed complex.

**Comment by saschanaz** [comment]
> A parser would need as much effort to parse `<aside>` than to find and parse a `<script type="application/json">`

It's not about efficiency, it's more about splitting - creating/opening panels at runtime while complex mapping at build time.

That said, there is no blocker here, let's :shipit: and see how it goes.

**Pull request merged**